network/ServerNetworking.cpp

```diff
@@ -30,11 +30,35 @@ namespace {
 class DiscoveryServer {
 public:
     DiscoveryServer(boost::asio::io_context& io_context) :
-        m_socket(io_context, udp::endpoint(udp::v6(), Networking::DiscoveryPort()))
+        m_service(io_context),
+        m_socket(io_context)
     { Listen(); }
 
 private:
     void Listen() {
+        // use a dual stack (ipv6 + ipv4) socket
+        udp::endpoint discovery_endpoint(udp::v6(), Networking::DiscoveryPort());
+
+        if (GetOptionsDB().Get<bool>("singleplayer")) {
+            // when hosting a single player game only accept connections from
+            // the localhost via the loopback interface instead of the any
+            // interface.
+            // This should prevent unnecessary triggering of Desktop Firewalls as
+            // reported by various users when running single player games.
+            discovery_endpoint.address(boost::asio::ip::address_v4::loopback());
```

Comment:
IPV4 loopback works with IPV6 endpoints?

Reply:
It works same way as `message_endpoint`.

Reply:
That doesn't answer the question, which is whether you can use an IPV4 loopback with an IPV6 endpoint.

Reply:
Looks like IPv6 is actually dual-stack endpoint.

Reply:
In that it works for both when both are present but also requires both to be present to work at all?

Reply:
Yes, it looks like it opens IPv6 port and allows IPv4 connections/packages be accepted here.